shared/sharedtest/util.go

```diff
@@ -25,28 +28,104 @@ type Instance interface {
 }
 
 type aeInstance struct {
-	Instance
+	// Google Cloud Datastore emulator
+	gcd      *exec.Cmd
+	hostPort string
+}
+
+func (i aeInstance) Close() error {
+	shared.Clients.Close()
+	return i.stop()
 }
 
 func (i aeInstance) NewRequest(method, urlStr string, body io.Reader) (*http.Request, error) {
-	req, err := i.Instance.NewRequest(method, urlStr, body)
-	if err != nil {
-		return req, err
+	req := httptest.NewRequest(method, urlStr, body)
+	return req.WithContext(ctxWithNilLogger(context.Background())), nil
+}
+
+func (i *aeInstance) start(stronglyConsistentDatastore bool) error {
+	consistency := "1.0"
+	if !stronglyConsistentDatastore {
+		consistency = "0.5"
+	}
+	// Project ID isn't important as long as it's valid.
+	project := "test-app"
+	// TODO(Hexcles): Check the port is available.
+	i.hostPort = "127.0.0.1:63842"
+	i.gcd = exec.Command("gcloud", "beta", "emulators", "datastore", "start",
+		"--no-store-on-disk",
+		"--consistency="+consistency,
+		"--project="+project,
+		"--host-port="+i.hostPort)
+	if err := i.gcd.Start(); err != nil {
+		return err
+	}
+
+	started := make(chan bool)
+	go func() {
+		for {
+			res, err := http.Get("http://" + i.hostPort)
+			if err == nil {
+				res.Body.Close()
+				if res.StatusCode == http.StatusOK {
+					started <- true
+					return
+				}
+			}
+			time.Sleep(time.Millisecond * 100)
+		}
+	}()
+	select {
+	case <-started:
+		break
+	case <-time.After(time.Second * 10):
+		i.stop()
+		return errors.New("timed out starting Datastore emulator")
+	}
+
+	os.Setenv("DATASTORE_PROJECT_ID", project)
```

Comment:
This seems to be a `gcloud beta emulators datastore env-init` step before this

Reply:
You're correct that these two `Setenv` are equivalent to `env-init`, but we can't use `env-init` here.

`env-init` simply prints out the shell commands to set these two env vars. And one needs to execute them in the shell, e.g. `$(gcloud beta emulators datastore env-init)`, but we are not in a shell here, and starting a new shell to execute these two commands won't affect the current process (and its future subprocesses).